ADAL/src/ADAuthenticationError.m

```diff
@@ -243,5 +246,45 @@ + (ADAuthenticationError*)errorFromNonHttpsRedirect:(NSUUID *)correlationId
                                                  correlationId:correlationId];
 }
 
++ (ADAuthenticationError *)keychainErrorFromOperation:(NSString *)operation
+                                               status:(OSStatus)status
+                                        correlationId:(NSUUID *)correlationId
+{
+    NSString* details = [NSString stringWithFormat:@"Keychain failed during \"%@\" operation", operation];
+    
+    return [self errorWithDomainInternal:ADKeychainErrorDomain
+                                    code:status
+                       protocolErrorCode:nil
+                            errorDetails:details
+                           correlationId:correlationId
+                                userInfo:nil];
+}
+
++ (ADAuthenticationError *)HTTPErrorCode:(NSInteger)code
+                                    body:(NSString *)body
+                           correlationId:(NSUUID *)correlationId
+{
+    return [self errorWithDomainInternal:ADHTTPErrorCodeDomain
+                                    code:code
+                       protocolErrorCode:nil
+                            errorDetails:body
+                           correlationId:correlationId
+                                userInfo:nil];
+}
+
++ (ADAuthenticationError *)OAuthServerError:(NSString *)protocolCode
+                                description:(NSString *)description
+                                       code:(NSInteger)code
+                              correlationId:(NSUUID *)correlationId
+{
+    return [self errorWithDomainInternal:ADOAuthServerErrorDomain
+                                    code:code
+                       protocolErrorCode:protocolCode
+                            errorDetails:description
+                           correlationId:correlationId
+                                userInfo:nil];
```

Comment:
This is an authentication error, not a log statement, and all of those details are accessible as properties already. userInfo is typically for extra information, which we'll eventually use for chaining errors